File: app/src/main/java/com/topjohnwu/magisk/database/RepoDatabaseHelper.java
package com.topjohnwu.magisk.database;

import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

import com.topjohnwu.magisk.Config;
import com.topjohnwu.magisk.container.Repo;

import java.util.HashSet;
import java.util.Set;

public class RepoDatabaseHelper extends SQLiteOpenHelper {

    private static final int DATABASE_VER = 5;
    private static final String TABLE_NAME = "repos";

    private SQLiteDatabase mDb;

    public RepoDatabaseHelper(Context context) {
        super(context, "repo.db", null, DATABASE_VER);
        mDb = getWritableDatabase();
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        onUpgrade(db, 0, DATABASE_VER);
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        if (oldVersion != newVersion) {
            // Nuke old DB and create new table
            db.execSQL("DROP TABLE IF EXISTS " + TABLE_NAME);
            db.execSQL(
                    "CREATE TABLE IF NOT EXISTS " + TABLE_NAME + " " +
                            "(id TEXT, name TEXT, version TEXT, versionCode INT, " +
                            "author TEXT, description TEXT, last_update INT, PRIMARY KEY(id))");
            Config.remove(Config.Key.ETAG_KEY);
        }
    }

    @Override
    public void onDowngrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        onUpgrade(db, 0, DATABASE_VER);
    }

    public void clearRepo() {
        mDb.delete(TABLE_NAME, null, null);
    }


    public void removeRepo(String id) {
        mDb.delete(TABLE_NAME, "id=?", new String[] { id });
    }

    public void removeRepo(Repo repo) {
        removeRepo(repo.getId());
    }

    public void removeRepo(Iterable<String> list) {
        for (String id : list) {
            if (id == null) continue;
            mDb.delete(TABLE_NAME, "id=?", new String[] { id });
        }
    }

    public void addRepo(Repo repo) {
        mDb.replace(TABLE_NAME, null, repo.getContentValues());
    }

    public Repo getRepo(String id) {
        try (Cursor c = mDb.query(TABLE_NAME, null, "id=?", new String[] { id }, null, null, null)) {
            if (c.moveToNext()) {
                return new Repo(c);
            }
        }
        return null;
    }

    public Cursor getRawCursor() {
        return mDb.query(TABLE_NAME, null, null, null, null, null, null);
    }

    public Cursor getRepoCursor() {
        String orderBy = null;
        switch ((int) Config.get(Config.Key.REPO_ORDER)) {
            case Config.Value.ORDER_NAME:
                orderBy = "name COLLATE NOCASE";
                break;
            case Config.Value.ORDER_DATE:
                orderBy = "last_update DESC";
        }
        return mDb.query(TABLE_NAME, null, null, null, null, null, orderBy);
    }

    public Set<String> getRepoIDSet() {
        HashSet<String> set = new HashSet<>(300);
        try (Cursor c = mDb.query(TABLE_NAME, null, null, null, null, null, null)) {
            while (c.moveToNext()) {
                set.add(c.getString(c.getColumnIndex("id")));
            }
        }
        return set;
    }
}

File: app/src/main/java/com/topjohnwu/magisk/data/database/RepoDatabaseHelper.kt
package com.topjohnwu.magisk.data.database

import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import androidx.core.content.edit
import com.topjohnwu.magisk.Config
import com.topjohnwu.magisk.model.entity.Repo
import java.util.*

@Deprecated("")
class RepoDatabaseHelper
constructor(context: Context) : SQLiteOpenHelper(context, "repo.db", null, DATABASE_VER) {

    private val mDb: SQLiteDatabase = writableDatabase

    val rawCursor: Cursor
        @Deprecated("")
        get() = mDb.query(TABLE_NAME, null, null, null, null, null, null)

    val repoCursor: Cursor
        @Deprecated("")
        get() {
            var orderBy: String? = null
            when (Config.repoOrder) {
                Config.Value.ORDER_NAME -> orderBy = "name COLLATE NOCASE"
                Config.Value.ORDER_DATE -> orderBy = "last_update DESC"
            }
            return mDb.query(TABLE_NAME, null, null, null, null, null, orderBy)
        }

    val repoIDSet: Set<String>
        @Deprecated("")
        get() {
            val set = HashSet<String>(300)
            mDb.query(TABLE_NAME, null, null, null, null, null, null).use { c ->
                while (c.moveToNext()) {
                    set.add(c.getString(c.getColumnIndex("id")))
                }
            }
            return set
        }

    override fun onCreate(db: SQLiteDatabase) {
        onUpgrade(db, 0, DATABASE_VER)
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        if (oldVersion != newVersion) {
            // Nuke old DB and create new table
            db.execSQL("DROP TABLE IF EXISTS $TABLE_NAME")
            db.execSQL(
                    "CREATE TABLE IF NOT EXISTS " + TABLE_NAME + " " +
                            "(id TEXT, name TEXT, version TEXT, versionCode INT, " +
                            "author TEXT, description TEXT, last_update INT, PRIMARY KEY(id))")
            Config.prefs.edit {
                remove(Config.Key.ETAG_KEY)
            }
        }
    }

    override fun onDowngrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        onUpgrade(db, 0, DATABASE_VER)
    }

    @Deprecated("")
    fun clearRepo() {
        mDb.delete(TABLE_NAME, null, null)
    }


    @Deprecated("")
    fun removeRepo(id: String) {
        mDb.delete(TABLE_NAME, "id=?", arrayOf(id))
    }

    @Deprecated("")
    fun removeRepo(repo: Repo) {
        removeRepo(repo.id)
    }

    @Deprecated("")
    fun removeRepo(list: Iterable<String>) {
        list.forEach {
            mDb.delete(TABLE_NAME, "id=?", arrayOf(it))
        }
    }

    @Deprecated("")
    fun addRepo(repo: Repo) {
        mDb.replace(TABLE_NAME, null, repo.contentValues)
    }

    @Deprecated("")
    fun getRepo(id: String): Repo? {
        mDb.query(TABLE_NAME, null, "id=?", arrayOf(id), null, null, null).use { c ->
            if (c.moveToNext()) {
                return Repo(c)
            }
        }
        return null
    }

    companion object {
        private val DATABASE_VER = 5
        private val TABLE_NAME = "repos"
    }
}
